### PR TITLE
Adds Nesterov momentum to SGD

### PR DESCRIPTION
**Pull request by abeleinin**
Added Nesterov momentum to SGD on recommendation from issue #46.

### CODE REVIEW

File: python/mlx/optimizers.py
Comment (awni) [inline python/mlx/optimizers.py:114]: Which version of Nesterov are you going for? The original in the paper or the PyTorch flavor? This doesn't look like the [PyTorch version](https://pytorch.org/docs/stable/generated/torch.optim.SGD.html) to me.

Also I'm just noticing this now, but I think our momentum is slightly non-standard in that we scale the gradient by `(1-self.momentum)`. Should we change that? (CC @angeloskath ).

Comment (angeloskath) [reply python/mlx/optimizers.py:114]: Agreed. Going directly with the PyTorch version of both would be the simplest.

File: python/mlx/optimizers.py
Comment (angeloskath) [inline python/mlx/optimizers.py:112]: This would need to move after the update of `v`.

**Review by angeloskath — CHANGES_REQUESTED** [review]
Thanks for adding this!

I would fix the Nesterov update and since you are at it feel free to add custom dampening (we have it fixed at `1-momentum` now) and maybe also weight decay.

**Comment by abeleinin** [comment]
Thank you for the feedback. I just updated the PR to match the implementation from PyTorch for Nesterov momentum, `dampening`, and `weight_decay`. Let me know your thoughts.

File: python/mlx/optimizers.py
Comment (awni) [inline python/mlx/optimizers.py:130]: I don't think this is correct. You don't want to update `v` in the nesterov case.

I think it should be something like:

```
v = self.momentum * v
if dampening > 0:
  v += (1 - dampening) * gradient
else:
  v += gradient

if self.nesterov
  update = gradient  + self.momentum * v
else:
  update = v
state["v"] = v
return parameter - self.learning_rate * update
```

**Review by awni — CHANGES_REQUESTED** [review]

File: python/mlx/optimizers.py
Comment (abeleinin) [reply python/mlx/optimizers.py:130]: My bad. I will update this.

Also, in PyTorch's SGD constructor, they have this check:
```
if nesterov and (momentum <= 0 or dampening != 0):
    raise ValueError("Nesterov momentum requires a momentum and zero dampening")
```
https://pytorch.org/docs/stable/_modules/torch/optim/sgd.html#SGD
Should this be added since I'm adding the `dampening` parameter?

**Comment by abeleinin** [comment]
> > Thank you for the feedback. I just updated the PR to match the implementation from PyTorch for Nesterov momentum, `dampening`, and `weight_decay`. Let me know your thoughts.
> 
> Linguistics nitpick: it should be "damping" (a decrease in the amplitude of an oscillation as a result of energy being drained from the system to overcome frictional or other resistive forces) rather than "dampening" (make slightly wet).

PyTorch's implementation uses the word `dampening`. I don't have a strong preference either way, so I'll let the reviewers make the final call.

File: python/mlx/optimizers.py
Comment (awni) [reply python/mlx/optimizers.py:130]: Yea, good call, I think it's worth adding.

**Comment by awni** [comment]
I'm cool with `dampening`

Screenshot 2023-12-08 at 8 23 34 PM

**Comment by abeleinin** [comment]
Let me know if you want me to change the documentation. I removed the `(1 - \mu)` term since it's not accurate anymore.

**Review by awni — APPROVED** [review]
Looks great, thanks!!

**Review by awni — APPROVED** [review]

**Comment by awni** [comment]
@angeloskath needs your approval before we can merge this

**Review by angeloskath — APPROVED** [review]
This is great! Thanks.

**Pull request merged**
